File: app/src/main/java/com/example/josycom/flowoverstack/repository/SearchRepository.java
package com.example.josycom.flowoverstack.repository;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;

import com.example.josycom.flowoverstack.model.QuestionsResponse;
import com.example.josycom.flowoverstack.model.SearchResponse;
import com.example.josycom.flowoverstack.network.ApiService;
import com.example.josycom.flowoverstack.network.RestApiClient;
import com.example.josycom.flowoverstack.util.StringConstants;
import com.example.josycom.flowoverstack.util.ThreadExecutor;

import org.jetbrains.annotations.NotNull;

import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchRepository {

    private MutableLiveData<SearchResponse> mResponse = new MutableLiveData<>();
    private ThreadExecutor threadExecutor = new ThreadExecutor();

    private void getQuestionsWithTextInTitle(String inTitle) {
        ApiService apiService = RestApiClient.getApiService(ApiService.class);
        mResponse.postValue(new SearchResponse(StringConstants.LOADING, null));
        Call<QuestionsResponse> call = apiService.getQuestionsWithTextInTitle(inTitle);
        call.enqueue(new Callback<QuestionsResponse>() {
            @Override
            public void onResponse(@NotNull Call<QuestionsResponse> call, @NotNull Response<QuestionsResponse> response) {
                QuestionsResponse questionsResponse = response.body();
                if (Objects.requireNonNull(questionsResponse).getItems().size() != 0) {
                    mResponse.setValue(new SearchResponse(StringConstants.LOADED, questionsResponse.getItems()));
                } else {
                    Log.d("SearchRepository", "No matching question");
                    mResponse.setValue(new SearchResponse(StringConstants.NO_MATCHING_RESULT, null));
                }
            }

            @Override
            public void onFailure(@NotNull Call<QuestionsResponse> call, @NotNull Throwable t) {
                t.printStackTrace();
                mResponse.setValue(new SearchResponse(StringConstants.FAILED, null));
            }
        });
    }

    public MutableLiveData<SearchResponse> getResponse(String inTitle) {
        threadExecutor.mExecutor.execute(() -> getQuestionsWithTextInTitle(inTitle));
        return mResponse;
    }
}

File: app/src/main/java/com/josycom/mayorjay/flowoverstack/data/SearchRepository.kt
package com.josycom.mayorjay.flowoverstack.data

import androidx.lifecycle.MutableLiveData
import com.josycom.mayorjay.flowoverstack.network.QuestionsResponse
import com.josycom.mayorjay.flowoverstack.network.SearchResponse
import com.josycom.mayorjay.flowoverstack.network.ApiService
import com.josycom.mayorjay.flowoverstack.util.AppConstants
import com.josycom.mayorjay.flowoverstack.util.ThreadExecutor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SearchRepository @Inject constructor(private val apiService: ApiService) {

    private val mResponse = MutableLiveData<SearchResponse>()

    private fun getQuestionsWithTextInTitle(inTitle: String, page: Int, pageSize: Int) {
        mResponse.postValue(SearchResponse(AppConstants.LOADING, null))
        val call = apiService.getQuestionsWithTextInTitle(inTitle, page, pageSize)
        call.enqueue(object : Callback<QuestionsResponse?> {
            override fun onResponse(call: Call<QuestionsResponse?>, response: Response<QuestionsResponse?>) {
                val questionsResponse = response.body()
                if (questionsResponse != null) {
                    if (questionsResponse.items.isNotEmpty()) {
                        mResponse.setValue(SearchResponse(AppConstants.LOADED, questionsResponse.items))
                    } else {
                        mResponse.setValue(SearchResponse(AppConstants.NO_MATCHING_RESULT, null))
                    }
                }
            }

            override fun onFailure(call: Call<QuestionsResponse?>, t: Throwable) {
                t.printStackTrace()
                mResponse.value = SearchResponse(AppConstants.FAILED, null)
            }
        })
    }

    fun getResponse(inTitle: String, page: Int, pageSize: Int): MutableLiveData<SearchResponse> {
        ThreadExecutor.mExecutor.execute { getQuestionsWithTextInTitle(inTitle, page, pageSize) }
        return mResponse
    }
}